logverification/app/appentry.go

```diff
@@ -0,0 +1,292 @@
+package app
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/datatrails/go-datatrails-common-api-gen/assets/v2/assets"
+	"github.com/datatrails/go-datatrails-common/logger"
+	"github.com/datatrails/go-datatrails-merklelog/massifs"
+	"github.com/datatrails/go-datatrails-merklelog/mmr"
+	"github.com/google/uuid"
+)
+
+/**
+ * AppEntry is the app provided data for a corresponding log entry.
+ * An AppEntry will derive fields used for log entry inclusion verification.
+ *
+ * The format for an MMR entry is the following:
+ *
+ * H( Domain | MMR Salt | Serialized Bytes)
+ *
+ * Where:
+ *   * Domain           - the hashing schema for the MMR Entry
+ *   * MMR Salt         - datatrails provided fields included in the MMR Entry (can be found in the corresponding Trie Value on the log)
+ *   * Serialized Bytes - app (customer) provided fields in the MMR Entry, serialized in a consistent way.
+ *
+ *
+ * The format for a Trie Entry is the following:
+ *
+ * ( Trie Key | Trie Value )
+ *
+ * Where the Trie Key is:
+ *
+ * H( Domain | LogId | AppId )
+ *
+ * And Trie Value is:
+ *
+ * ( Extra Bytes | IdTimestamp )
+ */
+
+const (
+	MMRSaltSize = 32
+
+	IDTimestapSizeBytes = 8
+)
+
+// MMREntryFields are the fields that when hashed result in the MMR Entry
+type MMREntryFields struct {
+
+	// Domain defines the hashing schema for the MMR Entry
+	Domain byte
+
+	// SerializedBytes are app (customer) provided fields in the MMR Entry, serialized in a consistent way.
+	SerializedBytes []byte
+}
+
+// AppEntry is the app provided data for a corresponding log entry.
+//
+// It contains key information for verifying inclusion of the corresponding log entry.
+//
+// NOTE: all fields are sourced from the app data, or derived from it.
+// NONE of the fields in an AppEntry are sourced from the log.
+type AppEntry struct {
+	// AppId is an identifier of the app committing the merkle log entry
+	AppId string
+
+	// LogId is a uuid in byte form of the specific log identifier
+	LogId []byte
+
+	// ExtraBytes are extrabytes provided by datatrails for the specific app
+	ExtraBytes []byte
+
+	// MMREntryFields used to determine the MMR Entry
+	MMREntryFields *MMREntryFields
+
+	// MerkleLogCommit used to define information about the log entry
+	MerkleLogCommit *assets.MerkleLogCommit
+
+	// MerkleLogConfirm used to define information about the log seal
+	MerkleLogConfirm *assets.MerkleLogConfirm
+}
```

Comment:
Its usually good practice to make all fields private i.e. begin with lowercase letter. This encourages good habits
If you have a good reason to make them public then add a comment

Reply:
Public because we want access to them, are you suggesting we should make them private then have getter methods?

Reply:
Genrally all fields of a struct should be private i.e. an opaque interface is preferred. The grpc plugin for protoc has recently introduced an opaque interface to the generated code.

So usually one makes fields private unless there is a requirement to access them. In that case add a comment as to why the field is public OR add a Getter method.

A getter method is preferred because this prevents a user from changing the value unexpectedly.

i.e a public field is read/writable by the user, a getter method is readonly and a private field is neither.

A nice comment as to why each field is read/write, readonly will explain why this is so.

Reply:
and call the gettermethod AppID() (not GetAppID())

Reply:
done!

this is much cleaner thank you :) 

have defined 3 interfaces as well that helps

Reply:
As an aside when marshalling JSON into a struct the fields must be public. In this case this makes sense because JSON statements are by their very nature public. 
Because we do the JSON marshalling thing a lot we have fallen into the habit of always making fields public even when this is sub-optimal.

Reply:
thats a fair point :)